In [15]:
!pip install keras

In [16]:
import tensorflow
from tensorflow.keras import datasets
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Flatten, Dropout, Dense

In [17]:
fashion = datasets.fashion_mnist

(X_train, y_train), (X_test, y_test) = fashion.load_data()

In [18]:
# Normalize the pixel from picture so the value was 0-1
X_train = X_train/255.0
X_test = X_test/255.0

## Prepare an ANN model with one hidden layer containing 128 nodes

In [19]:
model1 = tensorflow.keras.models.Sequential()
model1.add(Flatten(input_shape=[28, 28]))
model1.add(Dense(128, activation='relu'))
model1.add(Dense(10, activation='softmax'))

/usr/local/lib/python3.12/dist-packages/keras/src/layers/reshaping/flatten.py:37: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


In [36]:
# Compile the model using SGD as the optimizer
model1.compile(loss='sparse_categorical_crossentropy',
optimizer = 'SGD',
metrics = ['accuracy'])

In [21]:
model1.summary()

Model: "sequential_2"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ flatten_2 (Flatten)             │ (None, 784)            │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_4 (Dense)                 │ (None, 128)            │       100,480 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_5 (Dense)                 │ (None, 10)             │         1,290 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 101,770 (397.54 KB)

 Trainable params: 101,770 (397.54 KB)

 Non-trainable params: 0 (0.00 B)

In [22]:
model1.fit(X_train, y_train, epochs = 100,
           validation_split = 0.2,
           batch_size = 128)

Epoch 1/100
375/375 ━━━━━━━━━━━━━━━━━━━━ 5s 11ms/step - accuracy: 0.5452 - loss: 1.4687 - val_accuracy: 0.7429 - val_loss: 0.8047
Epoch 2/100
375/375 ━━━━━━━━━━━━━━━━━━━━ 3s 5ms/step - accuracy: 0.7525 - loss: 0.7686 - val_accuracy: 0.7807 - val_loss: 0.6706
Epoch 3/100
375/375 ━━━━━━━━━━━━━━━━━━━━ 3s 5ms/step - accuracy: 0.7868 - loss: 0.6578 - val_accuracy: 0.7974 - val_loss: 0.6081
Epoch 4/100
375/375 ━━━━━━━━━━━━━━━━━━━━ 2s 6ms/step - accuracy: 0.8059 - loss: 0.5988 - val_accuracy: 0.8066 - val_loss: 0.5726
Epoch 5/100
375/375 ━━━━━━━━━━━━━━━━━━━━ 3s 6ms/step - accuracy: 0.8133 - loss: 0.5672 - val_accuracy: 0.8133 - val_loss: 0.5493
Epoch 6/100
375/375 ━━━━━━━━━━━━━━━━━━━━ 2s 5ms/step - accuracy: 0.8209 - loss: 0.5443 - val_accuracy: 0.8213 - val_loss: 0.5275
Epoch 7/100
375/375 ━━━━━━━━━━━━━━━━━━━━ 3s 6ms/step - accuracy: 0.8269 - loss: 0.5157 - val_accuracy: 0.8256 - val_loss: 0.5114
Epoch 8/100
375/375 ━━━━━━━━━━━━━━━━━━━━ 3s 6ms/step - accuracy: 0.8296 - loss: 0.5120 - val_acc

In [23]:
model1.evaluate(X_test, y_test)

313/313 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.8665 - loss: 0.3754


[0.3778742551803589, 0.8654999732971191]

## Prepare a DNN model with two hidden layers, each consisting of 128 nodes

In [24]:
model2 = tensorflow.keras.models.Sequential()
model2.add(Flatten(input_shape=[28, 28]))
model2.add(Dense(128, activation = 'relu'))
model2.add(Dense(10, activation = 'softmax'))

In [37]:
# Similar to model 1, compile the model using SGD as the optimizer
model2.compile(loss='sparse_categorical_crossentropy',
optimizer = 'SGD',
metrics = ['accuracy'])

In [26]:
model2.summary()

Model: "sequential_3"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ flatten_3 (Flatten)             │ (None, 784)            │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_6 (Dense)                 │ (None, 128)            │       100,480 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_7 (Dense)                 │ (None, 10)             │         1,290 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 101,770 (397.54 KB)

 Trainable params: 101,770 (397.54 KB)

 Non-trainable params: 0 (0.00 B)

In [27]:
model2.fit(X_train, y_train, epochs = 100,
validation_split = 0.2,
batch_size = 128)

Epoch 1/100
375/375 ━━━━━━━━━━━━━━━━━━━━ 4s 10ms/step - accuracy: 0.5447 - loss: 1.4746 - val_accuracy: 0.7406 - val_loss: 0.7922
Epoch 2/100
375/375 ━━━━━━━━━━━━━━━━━━━━ 5s 10ms/step - accuracy: 0.7485 - loss: 0.7632 - val_accuracy: 0.7832 - val_loss: 0.6632
Epoch 3/100
375/375 ━━━━━━━━━━━━━━━━━━━━ 3s 5ms/step - accuracy: 0.7878 - loss: 0.6473 - val_accuracy: 0.7978 - val_loss: 0.6063
Epoch 4/100
375/375 ━━━━━━━━━━━━━━━━━━━━ 3s 5ms/step - accuracy: 0.8048 - loss: 0.5954 - val_accuracy: 0.8108 - val_loss: 0.5676
Epoch 5/100
375/375 ━━━━━━━━━━━━━━━━━━━━ 3s 5ms/step - accuracy: 0.8145 - loss: 0.5602 - val_accuracy: 0.8167 - val_loss: 0.5427
Epoch 6/100
375/375 ━━━━━━━━━━━━━━━━━━━━ 3s 8ms/step - accuracy: 0.8224 - loss: 0.5388 - val_accuracy: 0.8209 - val_loss: 0.5255
Epoch 7/100
375/375 ━━━━━━━━━━━━━━━━━━━━ 2s 6ms/step - accuracy: 0.8250 - loss: 0.5248 - val_accuracy: 0.8215 - val_loss: 0.5142
Epoch 8/100
375/375 ━━━━━━━━━━━━━━━━━━━━ 2s 5ms/step - accuracy: 0.8305 - loss: 0.5017 - val_ac

In [28]:
model2.evaluate(X_test,y_test)

313/313 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.8709 - loss: 0.3695


[0.37471649050712585, 0.8669999837875366]

## Prepare a DNN model with two hidden layers, each containing 128 nodes. This architecture is identical to model2

In [34]:
model3 = tensorflow.keras.models.Sequential()
model3.add(Flatten(input_shape=[28, 28]))
model3.add(Dense(128, activation = 'relu'))
model3.add(Dense(128, activation = 'relu'))
model3.add(Dense(10, activation = 'softmax'))

In [35]:
# Unlike model2, model3 uses the Adam optimizer instead of SGD.
model3.compile(loss='sparse_categorical_crossentropy',
optimizer = 'Adam',
metrics = ['accuracy'])

In [31]:
model3.summary()

Model: "sequential_4"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ flatten_4 (Flatten)             │ (None, 784)            │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_8 (Dense)                 │ (None, 128)            │       100,480 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_9 (Dense)                 │ (None, 128)            │        16,512 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_10 (Dense)                │ (None, 10)             │         1,290 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 118,282 (462.04 KB)

 Trainable params: 118,282 (462.04 KB)

 Non-trainable params: 0 (0.00 B)

In [32]:
model3.fit(X_train, y_train,
epochs = 100,
validation_split = 0.2,
batch_size = 128)

Epoch 1/100
375/375 ━━━━━━━━━━━━━━━━━━━━ 4s 7ms/step - accuracy: 0.7394 - loss: 0.7874 - val_accuracy: 0.8411 - val_loss: 0.4583
Epoch 2/100
375/375 ━━━━━━━━━━━━━━━━━━━━ 5s 6ms/step - accuracy: 0.8546 - loss: 0.4142 - val_accuracy: 0.8661 - val_loss: 0.3781
Epoch 3/100
375/375 ━━━━━━━━━━━━━━━━━━━━ 3s 6ms/step - accuracy: 0.8707 - loss: 0.3542 - val_accuracy: 0.8723 - val_loss: 0.3608
Epoch 4/100
375/375 ━━━━━━━━━━━━━━━━━━━━ 2s 6ms/step - accuracy: 0.8821 - loss: 0.3271 - val_accuracy: 0.8742 - val_loss: 0.3504
Epoch 5/100
375/375 ━━━━━━━━━━━━━━━━━━━━ 3s 8ms/step - accuracy: 0.8871 - loss: 0.3053 - val_accuracy: 0.8710 - val_loss: 0.3550
Epoch 6/100
375/375 ━━━━━━━━━━━━━━━━━━━━ 5s 7ms/step - accuracy: 0.8936 - loss: 0.2894 - val_accuracy: 0.8772 - val_loss: 0.3407
Epoch 7/100
375/375 ━━━━━━━━━━━━━━━━━━━━ 5s 6ms/step - accuracy: 0.8950 - loss: 0.2805 - val_accuracy: 0.8790 - val_loss: 0.3392
Epoch 8/100
375/375 ━━━━━━━━━━━━━━━━━━━━ 3s 9ms/step - accuracy: 0.9026 - loss: 0.2623 - val_accu

In [33]:
model3.evaluate(X_test, y_test)

313/313 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.8845 - loss: 0.9584


[0.939024031162262, 0.8866999745368958]

## For model 4, continue using a DNN with two hidden layers and the Adam optimizer, but add Dropout

In [39]:
model4 = tensorflow.keras.models.Sequential()
model4.add(Flatten(input_shape=[28, 28]))
model4.add(Dropout(0.3))
model4.add(Dense(128, activation = 'relu'))
model4.add(Dropout(0.3))
model4.add(Dense(128, activation = 'relu'))
model4.add(Dropout(0.3))
model4.add(Dense(10, activation = 'softmax'))

In [40]:
# Compile the model using the Adam optimizer
model4.compile(loss = 'sparse_categorical_crossentropy',
optimizer = 'Adam',
metrics = ['accuracy'])

In [41]:
model4.summary()

Model: "sequential_7"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ flatten_7 (Flatten)             │ (None, 784)            │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout_1 (Dropout)             │ (None, 784)            │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_15 (Dense)                │ (None, 128)            │       100,480 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout_2 (Dropout)             │ (None, 128)            │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_16 (Dense)                │ (None, 128)            │        16,512 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout_3 (Dropout)             │ (None, 128)            │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_17 (Dense)                │ (None, 10)             │         1,290 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 118,282 (462.04 KB)

 Trainable params: 118,282 (462.04 KB)

 Non-trainable params: 0 (0.00 B)

In [42]:
model4.fit(X_train, y_train, epochs = 100,
validation_split = 0.2,
batch_size = 128)

Epoch 1/100
375/375 ━━━━━━━━━━━━━━━━━━━━ 5s 11ms/step - accuracy: 0.5920 - loss: 1.1194 - val_accuracy: 0.8245 - val_loss: 0.4850
Epoch 2/100
375/375 ━━━━━━━━━━━━━━━━━━━━ 3s 7ms/step - accuracy: 0.7951 - loss: 0.5674 - val_accuracy: 0.8450 - val_loss: 0.4207
Epoch 3/100
375/375 ━━━━━━━━━━━━━━━━━━━━ 5s 7ms/step - accuracy: 0.8167 - loss: 0.5027 - val_accuracy: 0.8572 - val_loss: 0.3924
Epoch 4/100
375/375 ━━━━━━━━━━━━━━━━━━━━ 5s 7ms/step - accuracy: 0.8292 - loss: 0.4687 - val_accuracy: 0.8583 - val_loss: 0.3837
Epoch 5/100
375/375 ━━━━━━━━━━━━━━━━━━━━ 5s 7ms/step - accuracy: 0.8281 - loss: 0.4602 - val_accuracy: 0.8614 - val_loss: 0.3756
Epoch 6/100
375/375 ━━━━━━━━━━━━━━━━━━━━ 6s 10ms/step - accuracy: 0.8386 - loss: 0.4414 - val_accuracy: 0.8635 - val_loss: 0.3665
Epoch 7/100
375/375 ━━━━━━━━━━━━━━━━━━━━ 4s 7ms/step - accuracy: 0.8461 - loss: 0.4262 - val_accuracy: 0.8685 - val_loss: 0.3579
Epoch 8/100
375/375 ━━━━━━━━━━━━━━━━━━━━ 4s 11ms/step - accuracy: 0.8459 - loss: 0.4189 - val_a

In [43]:
model4.evaluate(X_test, y_test)

313/313 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.8862 - loss: 0.3122


[0.3135436177253723, 0.8877999782562256]

### Conclusion:
 - Model 1 (1 hidden layer, SGD, accuracy 86.7%)
A simple model, reasonably effective but limited in capturing complex patterns.

- Model 2 (2 hidden layers, SGD, accuracy 87.9%)
Adding a hidden layer increases representational capacity, improving accuracy by ~1%.

- Model 3 (2 hidden layers, Adam, accuracy 88.0%)
The Adam optimizer accelerates convergence and yields slightly better results than SGD.

- Model 4 (2 hidden layers + Dropout, Adam, accuracy 88.7%)
Dropout helps prevent overfitting and achieves the highest accuracy among all models.

Using a Sequential model, additional hidden layers can be included simply by stacking them as needed. As the architecture and optimization techniques (e.g., Adam + Dropout) become more complex, the model achieves better performance in Fashion-MNIST classification